integration/release_upgrade_linux_test.go

```diff
@@ -0,0 +1,364 @@
+	require.NoError(t, err)
+	return service
+}
+
+// criImageService returns cri.ImageManagerService based on the grpc address.
+func (p *ctrdProc) criImageService(t *testing.T) cri.ImageManagerService {
+	service, err := remote.NewImageService(p.grpcAddress(), 1*time.Minute)
+	require.NoError(t, err)
+	return service
+}
+
+// newCtrdProc is to start containerd process.
+func newCtrdProc(t *testing.T, ctrdBin string, ctrdWorkDir string) *ctrdProc {
+	p := &ctrdProc{workDir: ctrdWorkDir}
+
+	var args []string
+	args = append(args, "--root", p.rootPath())
+	args = append(args, "--state", p.statePath())
+	args = append(args, "--address", p.grpcAddress())
+	args = append(args, "--config", p.configPath())
+	args = append(args, "--log-level", "debug")
+
+	f, err := os.OpenFile(p.logPath(), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0600)
+	require.NoError(t, err, "open log file %s", p.logPath())
+	t.Cleanup(func() { f.Close() })
+
+	cmd := exec.Command(ctrdBin, args...)
+	cmd.Stdout = f
+	cmd.Stderr = f
+	cmd.SysProcAttr = &syscall.SysProcAttr{Pdeathsig: syscall.SIGKILL}
+
+	p.cmd = cmd
+	require.NoError(t, p.cmd.Start(), "start containerd(%s)", ctrdBin)
+	return p
+}
+
+// ctrdProc is used to control the containerd process's lifecycle.
+type ctrdProc struct {
+	// workDir has the following layout:
+	//
+	// - root  (dir)
+	// - state (dir)
+	// - containerd.sock (sock file)
+	// - config.toml (toml file, required)
+	// - containerd.log (log file, always open with O_APPEND)
+	workDir string
+	cmd     *exec.Cmd
+}
+
+// kill is to send the signal to containerd process.
+func (p *ctrdProc) kill(sig syscall.Signal) error {
+	return p.cmd.Process.Signal(sig)
+}
+
+// wait is to wait for exit event of containerd process.
+func (p *ctrdProc) wait() error {
+	return p.cmd.Wait()
+}
+
+// grpcAddress is to return containerd's address.
+func (p *ctrdProc) grpcAddress() string {
+	return filepath.Join(p.workDir, "containerd.sock")
+}
+
+// configPath is to return containerd's config file.
+func (p *ctrdProc) configPath() string {
+	return filepath.Join(p.workDir, "config.toml")
+}
+
+// rootPath is to return containerd's root path.
+func (p *ctrdProc) rootPath() string {
+	return filepath.Join(p.workDir, "root")
+}
+
+// statePath is to return containerd's state path.
+func (p *ctrdProc) statePath() string {
+	return filepath.Join(p.workDir, "state")
+}
+
+// logPath is to return containerd's log path.
+func (p *ctrdProc) logPath() string {
+	return filepath.Join(p.workDir, "containerd.log")
+}
+
+// isReady checks the containerd is ready or not.
+func (p *ctrdProc) isReady() error {
+	var (
+		service     cri.RuntimeService
+		err         error
+		ticker      = time.NewTicker(1 * time.Second)
+		ctx, cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+	)
+	defer func() {
+		cancel()
+		ticker.Stop()
+	}()
+
+	for {
+		select {
+		case <-ticker.C:
+			service, err = remote.NewRuntimeService(p.grpcAddress(), 5*time.Second)
```

Comment:
Should we query the introspection service and look for plugin status there instead?  That'd give a more-complete picture of whether containerd is up.

Reply:
The introspection service just shows that plugin has been registered. But it doesn't mean that the CRI plugin is ready to serve. The CRI plugin might take time to restore existing running containers. So, we have to use CRI client to check here.